In [1]:
import pandas as pd

In [2]:
horse_df = pd.read_csv(r"C:\Users\HP\Desktop\Data_Science\ML\Horse_ensemble\horse.csv")

horse_df.head()

surgery    age  hospital_number  rectal_temp  pulse  respiratory_rate  \
0      no  adult           530101         38.5   66.0              28.0   
1     yes  adult           534817         39.2   88.0              20.0   
2      no  adult           530334         38.3   40.0              24.0   
3     yes  young          5290409         39.1  164.0              84.0   
4      no  adult           530255         37.3  104.0              35.0   

  temp_of_extremities peripheral_pulse mucous_membrane capillary_refill_time  \
0                cool          reduced             NaN            more_3_sec   
1                 NaN              NaN   pale_cyanotic            less_3_sec   
2              normal           normal       pale_pink            less_3_sec   
3                cold           normal   dark_cyanotic            more_3_sec   
4                 NaN              NaN   dark_cyanotic            more_3_sec   

   ... packed_cell_volume total_protein abdomo_appearance abdomo_protein  \
0  ...               45.0           8.4               NaN            NaN   
1  ...               50.0          85.0            cloudy            2.0   
2  ...               33.0           6.7               NaN            NaN   
3  ...               48.0           7.2     serosanguious            5.3   
4  ...               74.0           7.4               NaN            NaN   

      outcome  surgical_lesion lesion_1 lesion_2  lesion_3  cp_data  
0        died               no    11300        0         0       no  
1  euthanized               no     2208        0         0       no  
2       lived               no        0        0         0      yes  
3        died              yes     2208        0         0      yes  
4        died               no     4300        0         0       no  

[5 rows x 28 columns]

In [3]:
horse_df.isnull().sum()

surgery                    0
age                        0
hospital_number            0
rectal_temp               60
pulse                     24
respiratory_rate          58
temp_of_extremities       56
peripheral_pulse          69
mucous_membrane           47
capillary_refill_time     32
pain                      55
peristalsis               44
abdominal_distention      56
nasogastric_tube         104
nasogastric_reflux       106
nasogastric_reflux_ph    246
rectal_exam_feces        102
abdomen                  118
packed_cell_volume        29
total_protein             33
abdomo_appearance        165
abdomo_protein           198
outcome                    0
surgical_lesion            0
lesion_1                   0
lesion_2                   0
lesion_3                   0
cp_data                    0
dtype: int64

In [4]:
horse_df.dtypes

surgery                   object
age                       object
hospital_number            int64
rectal_temp              float64
pulse                    float64
respiratory_rate         float64
temp_of_extremities       object
peripheral_pulse          object
mucous_membrane           object
capillary_refill_time     object
pain                      object
peristalsis               object
abdominal_distention      object
nasogastric_tube          object
nasogastric_reflux        object
nasogastric_reflux_ph    float64
rectal_exam_feces         object
abdomen                   object
packed_cell_volume       float64
total_protein            float64
abdomo_appearance         object
abdomo_protein           float64
outcome                   object
surgical_lesion           object
lesion_1                   int64
lesion_2                   int64
lesion_3                   int64
cp_data                   object
dtype: object

In [5]:
horse_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   surgery                299 non-null    object 
 1   age                    299 non-null    object 
 2   hospital_number        299 non-null    int64  
 3   rectal_temp            239 non-null    float64
 4   pulse                  275 non-null    float64
 5   respiratory_rate       241 non-null    float64
 6   temp_of_extremities    243 non-null    object 
 7   peripheral_pulse       230 non-null    object 
 8   mucous_membrane        252 non-null    object 
 9   capillary_refill_time  267 non-null    object 
 10  pain                   244 non-null    object 
 11  peristalsis            255 non-null    object 
 12  abdominal_distention   243 non-null    object 
 13  nasogastric_tube       195 non-null    object 
 14  nasogastric_reflux     193 non-null    object 
 15  nasoga

In [6]:
horse_df.shape

(299, 28)

In [7]:
horse_df["outcome"].value_counts()

lived         178
died           77
euthanized     44
Name: outcome, dtype: int64

# Treating With the null values

rectal_temp            239 non-null    float64
pulse                  275 non-null    float64
respiratory_rate       241 non-null    float64
nasogastric_reflux_ph  53 non-null     float64
packed_cell_volume     270 non-null    float64
total_protein          266 non-null    float64
abdomo_protein         101 non-null    float64

In [8]:
horse_df["rectal_temp"].fillna(horse_df["rectal_temp"].mean(),inplace = True)

In [9]:
horse_df["pulse"].fillna(horse_df["pulse"].mean(),inplace = True)

In [10]:
horse_df["respiratory_rate"].fillna(horse_df["respiratory_rate"].mean(),inplace = True)

In [11]:
horse_df["nasogastric_reflux_ph"].fillna(horse_df["nasogastric_reflux_ph"].mean(),inplace = True)

In [12]:
horse_df["packed_cell_volume"].fillna(horse_df["packed_cell_volume"].mean(),inplace = True)

In [13]:
horse_df["total_protein"].fillna(horse_df["total_protein"].mean(),inplace = True)

In [14]:
horse_df["abdomo_protein"].fillna(horse_df["abdomo_protein"].mean(),inplace = True)

# Another Way Of treating Null Values.

In [15]:
for i in horse_df.columns:
    if horse_df[i].hasnans:
        print(i)
        print(horse_df[i].dtypes)

temp_of_extremities
object
peripheral_pulse
object
mucous_membrane
object
capillary_refill_time
object
pain
object
peristalsis
object
abdominal_distention
object
nasogastric_tube
object
nasogastric_reflux
object
rectal_exam_feces
object
abdomen
object
abdomo_appearance
object


In [16]:
horse_df.head(20)

surgery    age  hospital_number  rectal_temp  pulse  respiratory_rate  \
0       no  adult           530101    38.500000   66.0         28.000000   
1      yes  adult           534817    39.200000   88.0         20.000000   
2       no  adult           530334    38.300000   40.0         24.000000   
3      yes  young          5290409    39.100000  164.0         84.000000   
4       no  adult           530255    37.300000  104.0         35.000000   
5       no  adult           528355    38.168619   72.0         30.460581   
6      yes  adult           526802    37.900000   48.0         16.000000   
7      yes  adult           529607    38.168619   60.0         30.460581   
8       no  adult           530051    38.168619   80.0         36.000000   
9       no  young          5299629    38.300000   90.0         30.460581   
10     yes  adult           528548    38.100000   66.0         12.000000   
11      no  adult           527927    39.100000   72.0         52.000000   
12     yes  adult           528031    37.200000   42.0         12.000000   
13      no  young          5291329    38.000000   92.0         28.000000   
14     yes  adult           534917    38.200000   76.0         28.000000   
15     yes  adult           530233    37.600000   96.0         48.000000   
16     yes  young          5301219    38.168619  128.0         36.000000   
17      no  adult           526639    37.500000   48.0         24.000000   
18     yes  adult          5290481    37.600000   64.0         21.000000   
19      no  adult           532110    39.400000  110.0         35.000000   

   temp_of_extremities peripheral_pulse mucous_membrane capillary_refill_time  \
0                 cool          reduced             NaN            more_3_sec   
1                  NaN              NaN   pale_cyanotic            less_3_sec   
2               normal           normal       pale_pink            less_3_sec   
3                 cold           normal   dark_cyanotic            more_3_sec   
4                  NaN              NaN   dark_cyanotic            more_3_sec   
5                 warm           normal       pale_pink            less_3_sec   
6               normal           normal     normal_pink            less_3_sec   
7                 cool              NaN             NaN            less_3_sec   
8                 cool           absent       pale_pink            less_3_sec   
9               normal              NaN     normal_pink            less_3_sec   
10                cool          reduced      bright_red            less_3_sec   
11                warm              NaN     bright_pink            less_3_sec   
12                warm           normal     normal_pink            less_3_sec   
13              normal           normal     bright_pink            less_3_sec   
14                cool           normal     normal_pink            less_3_sec   
15                cool           normal   pale_cyanotic            less_3_sec   
16                cool          reduced   pale_cyanotic            more_3_sec   
17                 NaN              NaN             NaN                   NaN   
18              normal           normal     bright_pink            less_3_sec   
19                cold          reduced   dark_cyanotic                   NaN   

    ... packed_cell_volume total_protein abdomo_appearance abdomo_protein  \
0   ...          45.000000      8.400000               NaN       3.039604   
1   ...          50.000000     85.000000            cloudy       2.000000   
2   ...          33.000000      6.700000               NaN       3.039604   
3   ...          48.000000      7.200000     serosanguious       5.300000   
4   ...          74.000000      7.400000               NaN       3.039604   
5   ...          46.307407     24.274436               NaN       3.039604   
6   ...          37.000000      7.000000               NaN       3.039604   
7   ...          44.000000      8.300000               NaN       3.039604   
8   ...    

temp_of_extremities
object
peripheral_pulse
object
mucous_membrane
object
capillary_refill_time
object
pain
object
peristalsis
object
abdominal_distention
object
nasogastric_tube
object
nasogastric_reflux
object
rectal_exam_feces
object
abdomen
object
abdomo_appearance
object

Since, we are not able to impute the categorical columns with their mode values,we will be using simple imputer.

In [17]:
# horse_df["temp_of_extremities"].fillna(horse_df["temp_of_extremities"].mode(),inplace = True)

In [18]:
#horse_df["peripheral_pulse"].fillna(horse_df["peripheral_pulse"].mode(),inplace = True)

In [19]:
#horse_df["mucous_membrane"].fillna(horse_df["mucous_membrane"].mode(),inplace = True)

In [20]:
#horse_df["temp_of_extremities"].fillna(horse_df["temp_of_extremities"].mode(),inplace = True)

In [21]:
#horse_df["temp_of_extremities"].fillna(horse_df["temp_of_extremities"].mode(),inplace = True)

In [22]:
#horse_df["temp_of_extremities"].fillna(horse_df["temp_of_extremities"].mode(),inplace = True)

In [23]:
#horse_df["temp_of_extremities"].fillna(horse_df["temp_of_extremities"].mode(),inplace = True)

OHE Encoder

In [24]:
# Encode the features

X = horse_df.drop("outcome" ,axis =1)
y = horse_df["outcome"]

In [25]:
X_encoded = pd.get_dummies(horse_df)
X_encoded

hospital_number  rectal_temp  pulse  respiratory_rate  \
0             530101    38.500000   66.0              28.0   
1             534817    39.200000   88.0              20.0   
2             530334    38.300000   40.0              24.0   
3            5290409    39.100000  164.0              84.0   
4             530255    37.300000  104.0              35.0   
..               ...          ...    ...               ...   
294           533886    38.168619  120.0              70.0   
295           527702    37.200000   72.0              24.0   
296           529386    37.500000   72.0              30.0   
297           530612    36.500000  100.0              24.0   
298           534618    37.200000   40.0              20.0   

     nasogastric_reflux_ph  packed_cell_volume  total_protein  abdomo_protein  \
0                 4.707547                45.0       8.400000        3.039604   
1                 4.707547                50.0      85.000000        2.000000   
2                 4.707547                33.0       6.700000        3.039604   
3                 5.000000                48.0       7.200000        5.300000   
4                 4.707547                74.0       7.400000        3.039604   
..                     ...                 ...            ...             ...   
294               4.707547                55.0      65.000000        3.039604   
295               4.707547                44.0      24.274436        3.300000   
296               4.707547                60.0       6.800000        3.039604   
297               4.707547                50.0       6.000000        3.400000   
298               4.707547                36.0      62.000000        1.000000   

     lesion_1  lesion_2  ...  abdomo_appearance_clear  \
0       11300         0  ...                        0   
1        2208         0  ...                        0   
2           0         0  ...                        0   
3        2208         0  ...                        0   
4        4300         0  ...                        0   
..        ...       ...  ...                      ...   
294      3205         0  ...                        0   
295      2208         0  ...                        0   
296      3205         0  ...                        0   
297      2208         0  ...                        0   
298      6112         0  ...                        1   

     abdomo_appearance_cloudy  abdomo_appearance_serosanguious  outcome_died  \
0                           0                                0             1   
1                           1                                0             0   
2                           0                                0             0   
3                           0                                1             1   
4                           0                                0             1   
..                        ...                              ...           ...   
294                         0                                0             0   
295                         0                                1             0   
296                         0                                0             1   
297                         0                                1             0   
298                         0                                0             0   

     outcome_euthanized  outcome_lived  surgical_lesion_no  \
0                     0              0                   1   
1                     1              0                   1   
2                     0              1                   1   
3                     0              0                   0   
4                     0              0                   1   
..                  ...            ...                 ...   
294                   1              0                   1   
295                   1              0                   0   
296                   0              0                   0   
297                  

In [26]:
X_encoded.shape

(299, 70)

In [27]:
X.shape

(299, 27)

In [28]:
from sklearn.preprocessing import LabelEncoder

horse_outcome_encoder =LabelEncoder()

y = horse_outcome_encoder.fit_transform(y)

print(y)

[0 1 2 0 0 2 2 0 1 2 2 2 2 0 2 0 0 2 2 2 2 2 2 2 2 2 2 2 0 2 0 1 2 2 2 1 1
 2 2 0 0 2 2 1 1 0 2 2 0 0 2 0 2 2 1 0 2 0 0 0 2 2 0 1 2 2 2 2 2 2 1 2 0 0
 0 1 2 2 0 2 0 2 2 0 2 2 0 1 2 2 2 0 2 0 2 2 2 1 2 2 2 1 2 2 0 2 2 2 1 0 0
 2 2 0 2 2 2 1 2 2 2 1 2 0 1 2 2 2 1 2 0 0 2 0 0 0 2 2 2 2 0 0 2 1 2 2 0 2
 2 2 0 2 2 2 2 2 1 2 2 2 0 2 1 2 1 2 2 2 1 2 0 2 0 0 2 2 0 2 1 0 2 0 2 2 2
 2 2 2 2 2 2 2 0 2 2 1 0 2 2 2 0 1 2 2 1 2 0 1 2 0 0 2 2 2 2 2 2 1 2 1 2 2
 2 2 2 0 2 1 2 2 0 2 2 2 0 2 2 1 1 0 2 0 2 2 0 2 0 2 2 1 0 0 2 2 0 1 0 2 0
 2 2 2 0 2 1 2 2 2 2 0 2 2 0 1 0 1 2 2 2 2 0 0 2 2 2 1 2 2 2 2 0 0 1 2 1 1
 0 2 1]


We can infer from the principle of a label encoder that

1. 0 Represents "died"
2. 1 Represents "euthanized"
3. 2 Represents "lived"

In [29]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_encoded , y ,test_size = 0.2, random_state =21)

In [30]:
print("Shape of X_train is : ", X_train.shape)
print("Shape of X_test is : ", X_test.shape)
print("Shape of y_train is : ", y_train.shape)
print("Shape of y_test is : ", y_test.shape)


Shape of X_train is :  (239, 70)
Shape of X_test is :  (60, 70)
Shape of y_train is :  (239,)
Shape of y_test is :  (60,)


In [31]:
from sklearn.impute import SimpleImputer
import numpy as np
imp = SimpleImputer(missing_values = np.nan, strategy = "most_frequent")
X_train = imp.fit_transform(X_train)
X_test = imp.fit_transform(X_test)

In [32]:
pd.DataFrame(X_train).isnull().sum()

0     0
1     0
2     0
3     0
4     0
     ..
65    0
66    0
67    0
68    0
69    0
Length: 70, dtype: int64

# Random Forest Classifier

In [33]:
from sklearn.ensemble import RandomForestClassifier


In [34]:
rfc_model = RandomForestClassifier(n_estimators = 150) # n_estimators refers to the number of decision tree used. 

In [35]:
rfc_model.fit(X_train,y_train)

RandomForestClassifier(n_estimators=150)

In [36]:
y_pred_test =rfc_model.predict(X_test)

In [37]:
from sklearn.metrics import accuracy_score
print("Accuracy score of the RFC model on test dataset is: ")
print(accuracy_score(y_test,y_pred_test))

Accuracy score of the RFC model on test dataset is: 
1.0


# Train Data Accuracy

In [38]:
y_pred_train =rfc_model.predict(X_train)

print("Accuracy score of the RFC model on test dataset is: ")
print(accuracy_score(y_train,y_pred_train))

Accuracy score of the RFC model on test dataset is: 
1.0


In [39]:
# Pruning the RFC
rfc_pruned_model = RandomForestClassifier(n_estimators=150,max_depth = 5) # Experimenting max depth parameter.


In [40]:
y_pred_test = rfc_pruned_model.fit(X_train,y_train)

In [41]:
y_pred_test =rfc_pruned_model.predict(X_test)

In [42]:
print("Accuracy score of the RFC model on test dataset is: ")
print(accuracy_score(y_test,y_pred_test))

Accuracy score of the RFC model on test dataset is: 
1.0


In [43]:
y_pred_train =rfc_pruned_model.predict(X_train)

In [44]:
print("Accuracy score of the RFC model on train dataset is: ")
print(accuracy_score(y_train,y_pred_train))

Accuracy score of the RFC model on train dataset is: 
1.0


# Adaboost Classifier

In [46]:
from sklearn.ensemble import AdaBoostClassifier

In [47]:
ada_model = AdaBoostClassifier(n_estimators = 150) # base_estimator = none means by default it will take decision tree.

ada_model.fit(X_train,y_train)

y_pred_ada_test =ada_model.predict(X_test)

y_pred_ada_train =ada_model.predict(X_train)


In [48]:
print("Accuracy score of the AdaBoost model on test dataset is: ")
print(accuracy_score(y_test,y_pred_ada_test))

Accuracy score of the AdaBoost model on test dataset is: 
1.0


In [50]:
print("Accuracy score of the AdaBoost model on train dataset is: ")
print(accuracy_score(y_train,y_pred_ada_train))

Accuracy score of the AdaBoost model on train dataset is: 
1.0


# End of RFC & AdaBoost Section.